examples/cfd/stokes_mgn/pi_inference.py

```diff
@@ -0,0 +1,561 @@
+            layers=[2, 128, 128, 128, 128, 3],
+            fourier_features=64,
+        ).to("cuda")
+
+        self.model = MdlsSymDNN(
+            input_keys=[Key("x"), Key("y")],
+            output_keys=[Key("u"), Key("v"), Key("p")],
+            layers=[2, 128, 128, 128, 128, 3],
+            fourier_features=64,
+        ).to(self.device)
+
+        self.node_pde = Stokes(nu=self.nu, dim=2)
+
+        self.nodes = self.node_pde.make_nodes() + [
+            self.model.make_node(name="flow_network", jit=False)
+        ]
+
+        # note: this example uses the Graph class from Modulus Sym to construct the
+        # computational graph. This allows you to leverage Modulus Sym's optimized
+        # derivative backend to compute the derivatives, along with other benefits like
+        # symbolic definition of PDEs and leveraging the PDEs from Modulus Sym's PDE
+        # module.
+        # For more details, refer: https://docs.nvidia.com/deeplearning/modulus/modulus-sym/api/modulus.sym.html#module-modulus.sym.graph
+        self.graph = Graph(
+            self.nodes,
+            [Key("x"), Key("y")],
+            [Key("u"), Key("v"), Key("p")],
+            func_arch=False,
+        ).to(
+            self.device
+        )  # For pure inference (no gradients)
+        self.graph_full = Graph(
+            self.nodes,
+            [Key("x"), Key("y")],
+            [
+                Key("u"),
+                Key("v"),
+                Key("p"),
+                Key("continuity"),
+                Key("momentum_x"),
+                Key("momentum_y"),
+            ],
+            func_arch=False,
+        ).to(self.device)
+
+        self.optimizer = torch.optim.Adam(self.model.parameters(), lr=0.001)
+
+    def parabolic_inflow(self, y, U_max=0.3):
+        u = 4 * U_max * y * (0.4 - y) / (0.4**2)
+        v = torch.zeros_like(y)
+        return u, v
+
+    def loss(self):
+        x_in, y_in = self.coords_inflow[:, 0:1], self.coords_inflow[:, 1:2]
+        results_inflow = self.graph_full.forward({"x": x_in, "y": y_in})
+        pred_u_in, pred_v_in = results_inflow["u"], results_inflow["v"]
+
+        x_no_slip, y_no_slip = self.coords_noslip[:, 0:1], self.coords_noslip[:, 1:2]
+        results_noslip = self.graph_full.forward({"x": x_no_slip, "y": y_no_slip})
+        pred_u_noslip, pred_v_noslip = results_noslip["u"], results_noslip["v"]
+
+        x_int, y_int = self.coords[:, 0:1], self.coords[:, 1:2]
+        results_int = self.graph_full.forward({"x": x_int, "y": y_int})
+        pred_mom_u, pred_mom_v, pred_cont = (
+            results_int["momentum_x"],
+            results_int["momentum_y"],
+            results_int["continuity"],
+        )
+        pred_u, pred_v, pred_p = results_int["u"], results_int["v"], results_int["p"]
+
+        u_in, v_in = self.parabolic_inflow(self.coords_inflow[:, 1:2])
+
+        # Compute losses
+        # data loss
+        loss_u = torch.mean((self.gnn_u - pred_u) ** 2)
```

Comment:
I don't understand this. This doesn't look like fine-tuning to me. It looks like a regression problem where you're training a PINN to satisfy the GNN outputs rather than the reference data. Why not train directly on the reference data and why is the GNN useful here?

Reply:
We cannot train on reference data as in the inference mode, ground truth is typically not available. Training on the GNN predictions along with the physics losses allow the predictions to be more accurate (This is made clear by the figure in the README, basically in the event where the GNN prediction is not accurate enough, the physics loss can help to guide the predictions to a more realistic one). I am also adding @mnabian to add more context if required as he was the primary author of this example.

Reply:
Resolving this as this is also being discussed here: https://github.com/NVIDIA/modulus/pull/275#discussion_r1432058257